src/org/ods/util/HelmStatusSimpleData.groovy

```diff
@@ -0,0 +1,186 @@
+            badTypes.addAll(expectedStringAttribsForRoot.second)
+            badTypes.addAll(expectedStringAttribsForInfo.second)
+            def att = "version"
+            if (!rootObject.containsKey(att)) {
+                missingKeys << att
+            } else if (!(rootObject[att] instanceof Integer)) {
+                badTypes << "${att}: expected Integer, found ${rootObject[att].getClass()}"
+            }
+            handleMissingKeysOrBadTypes(missingKeys, badTypes)
+            def resourcesObject = ensureMap(infoObject.resources, "info.resources")
+            // All resources are in an json object which organize resources by keys
+            // Examples are  "v1/Cluster", "v1/ConfigMap"  "v1/Deployment", "v1/Pod(related)"
+            // For these keys the map contains list of resources.
+            Map<String, List<String>> resourcesByKind = [:] .withDefault { [] }
+            for (entry in resourcesObject.entrySet()) {
+                def key = entry.key as String
+                def resourceList = ensureList(entry.value, "info.resources.${key}")
+                resourceList.eachWithIndex { resourceJsonObject, i ->
+                    def resourceContext = "info.resources.${key}.[${i}]"
+                    def resource = extractResource(resourceJsonObject, resourceContext)
+                    if (resource) {
+                        resourcesByKind[resource.first] << resource.second
+                    }
+                }
+            }
+            def hs = new HelmStatusSimpleData()
+            hs.with {
+                name = rootObject.name
+                version = rootObject.version as String
+                namespace = rootObject.namespace
+                status = infoObject.status
+                description = infoObject.description
+                lastDeployed = infoObject["last_deployed"]
+            }
+            hs.resourcesByKind = resourcesByKind.collectEntries { kind, names ->
+                [ kind, names.collect() ]
+            } as Map<String, List<String> >
+            return hs
+        } catch (IllegalArgumentException ex) {
+            throw new IllegalArgumentException(
+                "Unexpected helm status information in JSON at 'info': ${ex.message}")
+        }
+    }
+
+    @NonCPS
+    Map<String, List<String>> getResourcesByKind(List<String> kinds) {
+        return resourcesByKind.subMap(kinds)
+    }
+
+    @NonCPS
+    Map<String, Object> toMap() {
+        def result = [
+            name: name,
+            version: version,
+            namespace: namespace,
+            status: status,
+            description: description,
+            lastDeployed: lastDeployed,
+            resourcesByKind: resourcesByKind,
+        ]
+        return result
+    }
+
+    @NonCPS
+    String toString() {
+        return toMap().toMapString()
+    }
+
+    @NonCPS
+    private static Tuple2<String, String> extractResource(
```

Comment:
It's very difficult to understand this method. There must be a much simpler and clearer way to do this, which may even not need a method returning a tuple. This needs rewriting.

Reply:
The logic for this method is once again related to throwing exceptions with messages reporting what attributes are missing, see: https://github.com/opendevstack/ods-jenkins-shared-library/pull/1143#discussion_r1806269048